fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java

```diff
@@ -263,6 +271,19 @@ public class LuceneRecordCursor implements BaseCursor<IndexEntry> {
         closed = false;
     }
 
+    private void ensurePrimaryKeyIsInSort() {
+        SortField[] fields = sort.getSort();
+        if (fields.length < 2 || !LuceneIndexMaintainer.PRIMARY_KEY_SEARCH_NAME.equals(fields[1].getField())) {
+            SortField[] updatedFields = new SortField[fields.length + 1];
+            updatedFields[0] = fields[0];
+            updatedFields[1] = new SortField(LuceneIndexMaintainer.PRIMARY_KEY_SEARCH_NAME, SortField.Type.STRING, fields[0].getReverse());
```

Comment:
Why second and not last? Doesn't that break the contract of the sort, which so far has only been checked to have the partition key first. If the intention is just to break ties, it doesn't much matter where, I don't think.

Reply:
Could that cause it to flip-flop between partitions, if the second sort field is not the primary key? Could it cause data to be invisible?

Thinking about what it would look like in the pathological case, of having two partitions, entirely with the same partitioning field. I think some care is needed to make sure that you don't miss data.
Otherwise if you have
| primary key | value |
| -- | -- |
| a | 6 |
| b | 5 |
| c | 4 |
| d | 3 |

And that is in two partitions:

| primary key | value |
| -- | -- |
| c | 4 |
| d | 3 |

| primary key | value |
| -- | -- |
| a | 6 |
| b | 5 |

And you are searching descending by `value`. If you jump to the second partition, after continuing from the first, you need to make sure to clear out the `value` part of the continuation, but keep the partitioning field, otherwise it would end up skipping all of the second partition's value.

If it works as the last sort field, I think it would work in any position after the partitioning field.

This seems a bit tricky / complicated to get right; perhaps we could restrict in this version so that if the only sort provided is the partitioning field, we'll add the primary key, otherwise we will treat it as we would if it was not sorted by the partitioning field at all? Thoughts @MMcM

Reply:
Yes it does break the contract of the sort. However the intent was to convert the partitioning key in the sort to the actual **cross partition** unique key for a deterministic/correct list of matches, not for breaking ties.
Adding the primary key anywhere other than second in this case will not guarantee properly sorted results across partitions. Alternatively, (and perhaps the correct behavior), we can implicitly add the primary key **only** if the sort has the partition field alone.

Reply:
That makes sense. So perhaps it's best only thinking that it's sorted by partitioning key if it's that alone, plus an optional second primary key, which we otherwise add.